keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,578 @@
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. The main reason for this is that if a client has multiple code paths
+where it makes apply requests to the same object, but with different field sets,
+they must use different field manager names. If they use the same field manager
+name they will cause fields to be accidentally removed or disowned. This is a
+potential foot gun we would like to avoid.
+
+### Generated apply configuration types
+
+All fields present in an apply configuration become owned by the applier after
+when the apply request succeeds. Go structs contain zero valued fields which are
+included even if the user never explicitly sets the field. This means that all required 
+fields (must not be a pointer and must not be "omitempty") create fundamental soundness
+problem.
+
+In practice there are over 100 ints and bools that meet this criteria this in
+the Kubernetes API. Some examples:
+
+- [`ContainerPort.ContainerPort`](https://github.com/kubernetes/kubernetes/blob/6d76ece4d6e1ad01bad3e866279bc35065813ec7/staging/src/k8s.io/api/core/v1/types.go#L1836)
+- [`HorizontalPodAutoscalerSpec.MaxReplicas`](https://github.com/kubernetes/kubernetes/blob/6d76ece4d6e1ad01bad3e866279bc35065813ec7/staging/src/k8s.io/api/autoscaling/v1/types.go#L49)
+- [`ContainerStatus.Ready`](https://github.com/kubernetes/kubernetes/blob/6d76ece4d6e1ad01bad3e866279bc35065813ec7/staging/src/k8s.io/api/core/v1/types.go#L2470)
+
+While there are arguments to be made all of these fields are expected to cause
+problems in practice, there are some that clearly would, e.g. ContainerPort.
+
+Because of this we cannot use the existing go structs to represent apply
+configurations.
+
+<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
+Finalize which alternative to use based on developer feedback. See the
+[Alternatives](#alternatives) for a complete list, but are currently focusing on
+the two below alternatives. We are working with the Kubebuilder community to
+gather feedback on what developers prefer.
+<<[/UNRESOLVED]>>
+
+#### Alternative 1: Generated structs where all fields are pointers
+
+Example usage:
+
+```go
+import (
+  ptr "k8s.io/utils/pointer"
+)
+
+&appsv1apply.Deployment{
+  Name: ptr.StringPtr("nginx-deployment"),
+  Spec: &appsv1apply.DeploymentSpec{
+    Replicas: ptr.Int32Ptr(0),
+    Template: &v1apply.PodTemplate{
+      Spec: &v1apply.PodSpec{
+        Containers: []v1.Containers{
+          {
+            Name: ptr.StringPtr("nginx"),
+            Image: ptr.StringPtr("nginx:latest"),
+          },
+        }
+      },
+    },
+  },
+}
+```
+
+For this approach, developers need to use a library like
+https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
+to inline primitive literals.
+
+#### Alternative 2: Generated "builders"
```

Comment:
I like this option better, but I'm fine with both being created for comparison if you like.

Reply:
I expect this to be the most contentious part of the KEP, I started firmly in the "builders" (alternative 2) camp, now I'm a bit more on the fence, but we do have working implementations of both we are trying out: kubernetes/kubernetes#95988 (struct with pointers) and https://github.com/jpbetz/kubernetes/tree/apply-client-go-builders (builders).